File: src/main/java/buttondevteam/discordplugin/DiscordPlayerSender.java
package buttondevteam.discordplugin;

import buttondevteam.discordplugin.mcchat.MinecraftChatModule;
import buttondevteam.discordplugin.playerfaker.VCMDWrapper;
import discord4j.core.object.entity.MessageChannel;
import discord4j.core.object.entity.User;
import lombok.Getter;
import org.bukkit.entity.Player;
import org.mockito.Mockito;

import java.lang.reflect.Modifier;

public abstract class DiscordPlayerSender extends DiscordSenderBase implements IMCPlayer<DiscordPlayerSender> {

	protected Player player;
	private @Getter VCMDWrapper vanillaCmdListener;

	public DiscordPlayerSender(User user, MessageChannel channel, Player player, MinecraftChatModule module) {
		super(user, channel);
		this.player = player;
		vanillaCmdListener = new VCMDWrapper(VCMDWrapper.createListener(this, player, module));
	}

	@Override
	public void sendMessage(String message) {
		player.sendMessage(message);
		super.sendMessage(message);
	}

	@Override
	public void sendMessage(String[] messages) {
		player.sendMessage(messages);
		super.sendMessage(messages);
	}

	public static DiscordPlayerSender create(User user, MessageChannel channel, Player player, MinecraftChatModule module) {
		return Mockito.mock(DiscordPlayerSender.class, Mockito.withSettings().stubOnly().defaultAnswer(invocation -> {
			if (!Modifier.isAbstract(invocation.getMethod().getModifiers()))
				return invocation.callRealMethod();
			return invocation.getMethod().invoke(((DiscordPlayerSender) invocation.getMock()).player, invocation.getArguments());
		}).useConstructor(user, channel, player, module));
	}
}

File: src/main/scala/buttondevteam/discordplugin/mcchat/sender/DiscordPlayerSender.scala
package buttondevteam.discordplugin.mcchat.sender

import buttondevteam.discordplugin.mcchat.MinecraftChatModule
import discord4j.core.`object`.entity.User
import discord4j.core.`object`.entity.channel.MessageChannel
import org.bukkit.entity.Player
import org.mockito.Mockito
import org.mockito.invocation.InvocationOnMock

import java.lang.reflect.Modifier

object DiscordPlayerSender {
    def create(user: User, channel: MessageChannel, player: Player, module: MinecraftChatModule): DiscordPlayerSender =
        Mockito.mock(classOf[DiscordPlayerSender], Mockito.withSettings.stubOnly.defaultAnswer((invocation: InvocationOnMock) => {
            def foo(invocation: InvocationOnMock): AnyRef = {
                if (!Modifier.isAbstract(invocation.getMethod.getModifiers))
                    invocation.callRealMethod
                else
                    invocation.getMethod.invoke(invocation.getMock.asInstanceOf[DiscordPlayerSender].player, invocation.getArguments)
            }

            foo(invocation)
        }).useConstructor(user, channel, player, module))
}

abstract class DiscordPlayerSender(user: User, channel: MessageChannel, var player: Player, val module: Nothing) extends DiscordSenderBase(user, channel) with IMCPlayer[DiscordPlayerSender] with Player {

    override def getVanillaCmdListener = null

    override def sendMessage(message: String): Unit = {
        player.sendMessage(message)
        super.sendMessage(message)
    }

    override def sendMessage(messages: String*): Unit = {
        player.sendMessage(messages: _*)
        super.sendMessage(messages: _*)
    }
}